src/main/java/ch/uzh/ifi/hase/soprafs24/service/GameService.java

```diff
@@ -318,10 +322,31 @@ private void runActiveTurn(Game currentGame, Integer cardId) throws InterruptedE
     }
 
     public void handleInactivePlayer(Integer gameId) {
-        if (isGameInCalc(gameId)) return;
+        if (isGameInCalc(gameId)) {
+            deferredExecutionService.deferTask(gameId, () -> executeInactivePlayerLogic(gameId));
+        } else {
+            executeInactivePlayerLogic(gameId);
+        }
+    }
+
+    @EventListener
+    public void onGameCalculationStatusChange(InactiveRequestEvent event) {
+        if (deferredExecutionService.hasDeferredTask(event.getGameId())) {
+            deferredExecutionService.executeDeferredTask(event.getGameId());
+        }
+    }
+
+    private void executeInactivePlayerLogic(Integer gameId) {
         User inactivePlayer = UserContextHolder.getCurrentUser();
         Game currentGame = inMemoryGameRepository.findById(gameId);
-        if (currentGame != null && inactivePlayer.getUserId().equals(currentGame.getActivePlayer())) {
+
+        int historySize = currentGame.getHistory().size();
+
+        // We need to ensure that a correct selection at the very end of a turn does not in a wrongful change of Turn
+        if (historySize > 1 &&
+                (currentGame.getHistory().get(historySize - 1).getPicks().isEmpty()
+                        && currentGame.getHistory().get(historySize - 2).getUserId() == currentGame.getActivePlayer())) {return;}
+        if (inactivePlayer.getUserId().equals(currentGame.getActivePlayer())) {
             setCardsFaceDown(currentGame, currentGame.getHistory().get(currentGame.getHistory().size()-1));
             initiateNewTurn(currentGame, false);
```

Comment:
would the UserContextholder catche the user who initiated this inactive user method,

or (in case the new turn is initiated before this method) is it possible that userContextholder is updated if next player started to pick some cards?

Reply:
Since we only handle the deferred requests for the run turn (turning a card in the FE), this user is always the same. This deference can only happed for one player. player 1 turn, player 1 picks card, player 1 is inactive. As soon as the player sends the inactive req they cannot do anything anymore (the state of the game changes). no other player can turn a card or send an inactive req